### PR TITLE
store: use store.access to disable access to snap store

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This configuration value allows users to disable snapd's access to the snap store. This can be toggled like so:
```
# snap unset system store.access
# snap search mattermost-desktop
Name                        Version  Publisher      Notes  Summary
mattermost-desktop          5.4.0    snapcrafters✪  -      Open source, private cloud Slack-alternative
mattermost-desktop-moon127  5.4.0    moon127        -
# snap set system store.access=offline
# snap search mattermost-desktop
error: store is offline, use 'snap set system store.access=online' to go online
```

This PR uses `store.access` to disallow access to the snap store via the methods on `store.Store`. Additionally, `snap-repair` is modified to respect the `store.access` configuration value.

### CODE REVIEW

File: overlord/storecontext/context.go [hunk 1]
```diff
@@ -247,6 +257,22 @@ func (sc *storeContext) ProxyStoreParams(defaultURL *url.URL) (proxyStoreID stri
 	return "", defaultURL, nil
 }
 
+func (sc *storeContext) StoreAccess() (string, error) {
+	sc.state.Lock()
+	defer sc.state.Unlock()
+
+	access, err := sc.storeAccessQuerier.StoreAccess()
+	if err != nil && !errors.Is(err, state.ErrNoState) {
```
Comment (andrewphelpsj) [inline overlord/storecontext/context.go:265]: What is the difference between the usage of `config.NoOptionError` and `state.ErrNoState`?

File: store/store.go [hunk 2]
```diff
@@ -709,6 +709,16 @@ func (s *Store) retryRequestDecodeJSON(ctx context.Context, reqOptions *requestO
 
 // doRequest does an authenticated request to the store handling a potential macaroon refresh required if needed
 func (s *Store) doRequest(ctx context.Context, client *http.Client, reqOptions *requestOptions, user *auth.UserState) (*http.Response, error) {
+	// this gets used in retryRequestDecodeJSON, meaning we'll quickly retry
+	// this a few times. a potential solution is extending httputil.RetryRequest
+	// to short circuit on a specific set of errors?
+	//
+	// httputil.ShouldRetryError already does something like this. maybe add to
+	// that function a check for store.ErrStoreOffline?
```
Comment (andrewphelpsj) [inline store/store.go:717]: That can't actually be done, at least if the error lives in `store` because it would cause a cyclical import.

File: overlord/storecontext/context.go (hunk 1 above)
Comment (pedronis) [reply overlord/storecontext/context.go:265]: the former is only used by APIs directly related to the configstate, When we abstract away from those we tend not to use it

File: store/store.go (hunk 2 above)
Comment (pedronis) [reply store/store.go:717]: mmh,  doesn't httputil.ShouldRetryError(store.ErrStoreOffline) return false anyway at this point? that's at least the default

**Review by pedronis — COMMENTED** [review]
thank you, did a first pass

File: overlord/configstate/configcore/store.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+func handleStoreAccess(tr RunTransaction, opts *fsOnlyContext) error {
+	storeAccess, err := coreCfg(tr, "store.access")
+	if err != nil {
+		return err
+	}
+
+	// TODO: write something to disk somewhere for snap-repair to read from
+	// here?
+	_ = storeAccess
+
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/configstate/configcore/store.go:58]: Thank you. Is there a "right place" to write that file?

File: overlord/configstate/configcore/store.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func handleStoreAccess(_ sysconfig.Device, cfg ConfGetter, _ *fsOnlyContext) error {
+	storeAccess, err := coreCfg(cfg, "store.access")
+	if err != nil {
+		return err
+	}
+
+	// TODO: write something to disk somewhere for snap-repair to read from
+	// here?
```
Comment (pedronis) [inline overlord/configstate/configcore/store.go:58]: I see two options atm:

* a dedicate optional file for snapd to communicate options to snap-repair:  /var/lib/snapd/repair.json
* a file specific to store options but consumable by more things:  /var/lib/snapd/store.json

except for system-params which is meant for general use across tools that are not snapd, we usually go with per tool file(s)

File: store/store.go [hunk 5]
```diff
@@ -1587,6 +1640,25 @@ func (s *Store) snapConnCheck() ([]string, error) {
 	return hosts, nil
 }
 
+var ErrStoreOffline = errors.New("store is offline, use 'snap unset system store.access' to go online")
```
Comment (pedronis) [inline store/store.go:1643]: maybe "store is marked offline..." ?

File: overlord/configstate/configcore/store.go (hunk 4 above)
Comment (andrewphelpsj) [reply overlord/configstate/configcore/store.go:58]: I see that there is already a `/var/lib/snap/repair/repair.json`. I don't think that file would be work for this purpose. Do you think the tool-specific file should go inside of the `repair` directory? Or should it go at `/var/lib/snapd/repair.json`, like you suggested?

Here is usage of the existing file: https://github.com/snapcore/snapd/blob/619489d6ea9d4bb9a864b6bfbde5ba2fa77e7af8/cmd/snap-repair/runner.go#L838-L852, where `dirs.SnapRepairStateFile` is `/var/lib/snap/repair/repair.json`.

File: overlord/configstate/configcore/store.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+	switch storeAccess {
+	case "", "offline":
+		return nil
+	default:
+		return errInvalidStoreAccess
+	}
+}
+
+type RepairConfig struct {
```
Comment (andrewphelpsj) [inline overlord/configstate/configcore/store.go:56]: I have no clue if this is the right place for this type, or if we even want to use this type. We could just us a `map[string]any` if we want to eliminate it.

File: cmd/snap-repair/runner.go [hunk 7]
```diff
@@ -49,6 +49,7 @@ import (
 	"github.com/snapcore/snapd/httputil"
 	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/configcore"
```
Comment (pedronis) [inline cmd/snap-repair/runner.go:52]: in general configcore is designed to be built using build tags such that the fs-only subset of its functionality can be imported in most tools outside of snapd, while in general we don't want to import overlord manager bits in anything but snapd itself. Is still the case that usually we try to define APIs shared between configcore and the rest of the tools in sysconfig (you can take a look there), so something like RepairConfig would live in sysconfig. OTOH snap-repair is very special, as the fallback mechanism if snapd itself cannot operate we try to keep its import to a minimum, so for now would we would likely do is repeat the very small definition of an unexported repairConfig in two places. If it grows larger an approach would be to use a dedicated snap-repair/config package to share things between snap-repair and configcore.

File: store/store.go [hunk 8]
```diff
@@ -1642,7 +1642,7 @@ func (s *Store) snapConnCheck() ([]string, error) {
 
 var ErrStoreOffline = errors.New("store is marked offline, use 'snap unset system store.access' to go online")
 
-func (s *Store) checkStoreOnline() error {
+func (s *Store) checkStoreOffline() error {
```
Comment (pedronis) [inline store/store.go:1645]: this rename is strange, as I would expect such a function to error if the store is not offline, which is not what it does.

Comment (andrewphelpsj) [reply store/store.go:1645]: Yeah, I was debating it. I'll revert that.

File: overlord/storecontext/context.go [hunk 9]
```diff
@@ -69,9 +69,13 @@ type DeviceSessionRequestSigner interface {
...
 }
 
-type ProxyStoreer interface {
+type StoreOptions interface {
 	// ProxyStore returns the store assertion for the proxy store if one is set.
 	ProxyStore() (*asserts.Store, error)
+
+	// StoreAccess returns a string indicating whether the store should have
+	// network access or not
+	StoreAccess() (string, error)
```
Comment (MiguelPires) [inline overlord/storecontext/context.go:78]: Based on this interface alone, it's not clear what values this can return. If this returned a specific type based on an iota instead of a string, the values that can be returned would be constrained which would be clearer and less error prone

**Review by MiguelPires — COMMENTED** [review]
Some comments

File: overlord/storecontext/context.go (hunk 9 above)
Comment (andrewphelpsj) [reply overlord/storecontext/context.go:78]: Right now, we really treat it as a boolean. `offline`, or unset. Do you think a boolean would be better?

File: overlord/configstate/configcore/store_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+			"store.access": "offline",
+		},
+	})
+	c.Assert(err, IsNil)
+
+	f, err := os.Open(dirs.SnapRepairConfigFile)
+	c.Assert(err, IsNil)
+	defer f.Close()
+
+	var repairConfig configcore.RepairConfig
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/configstate/configcore/store_test.go:64]: Ugh, sorry. Missed this in the rename.

File: overlord/storecontext/context.go (hunk 9 above)
Comment (pedronis) [reply overlord/storecontext/context.go:78]: I'm fine with either introducing an enumeration or using StoreOffline() bool for the time being

Comment (andrewphelpsj) [reply overlord/storecontext/context.go]: Done, thanks!

File: store/store.go [hunk 11]
```diff
@@ -1587,7 +1601,30 @@ func (s *Store) snapConnCheck() ([]string, error) {
...
+	}
+
+	if access == "offline" {
+		return ErrStoreOffline
+	}
+
+	return nil
+}
+
 func (s *Store) ConnectivityCheck() (status map[string]bool, err error) {
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply store/store.go:1662]: I think that's a good idea. Otherwise, it'll be hard to make sure we don't miss any now or in the future

[the remark this answers is not in the document]
Comment (pedronis) [reply store/store.go:1662]: as we discussed separately one issue with that is that it might acess StoreOffline quite a bit which requires locking/unlocking the state

[the remark this answers is not in the document]
Comment (MiguelPires) [reply store/store.go:1662]: Yes, but I thought it would be fine since the locking, parsing the JSON and unlocking would still be orders of magnitude faster than the actual HTTP request so, in the end, it won't have a significant effect in latency. If it does makes a significant difference, could we cache this in the store configuration? By having the store register a callback that the handler in configcore would call to update the configuration's value or to invalidate the cached value so `checkOnline` has to re-read it.

[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply store/store.go:1662]: @pedronis do you know of a good way to test out the performance impact of locking/unlocking the state for each HTTP request that comes from the store?

Also, as written right now, all public methods on the store that do anything over the network will have to acquire the lock at some point. If that happen in the HTTP client or not, does that make a difference? Unless, do some store operations result in multiple HTTP requests?

[the remark this answers is not in the document]
Comment (pedronis) [reply store/store.go:1662]: I wouldn't agonize over this too much, most new methods that we would add medium term to the store we'll go through endpointURL, you might leave a TODO somewhere in the code about this issue

[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply store/store.go:1662]: Sound good, will do.

File: overlord/configstate/configcore/store.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+// repairConfig is a set of configuration data that is consumed by the
+// snap-repair command. This struct is duplicated in cmd/snap-repair.
+type repairConfig struct {
+	// StoreOffline is true if the store is marked as offline and should not be
+	// accessed.
+	StoreOffline bool `json:"store-offline"`
+}
+
+func handleStoreAccess(_ sysconfig.Device, cfg ConfGetter, _ *fsOnlyContext) error {
```
Comment (pedronis) [inline overlord/configstate/configcore/store.go:60]: we can't ignore the context here, look how other handlers use it, it can specify an alternative rootdir that this code needs to respect

File: cmd/snap-repair/runner.go
Comment (pedronis) [inline cmd/snap-repair/runner.go:1]: we likely want to split this change into a separate PR, we probably need to check how noisy are the snap-repair runs when offline and consider maybe bailing out at a higher level instead

**Review by pedronis — CHANGES_REQUESTED** [review]
thanks, couple comments, the configcore code need a fix/tweak. We might want to split out and do some manual testing of the repair runner changes to  see how noisy it gets

File: cmd/snap-repair/runner.go
Comment (andrewphelpsj) [reply cmd/snap-repair/runner.go:1]: Makes sense. What's the best way to test that out? Ubuntu core VM?

Comment (pedronis) [reply cmd/snap-repair/runner.go:1]: yes, ubuntu core vm seems the easiest

File: overlord/configstate/configcore/store.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/sysconfig"
+)
+
+func init() {
+	// add supported configuration of this module
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/configstate/configcore/store.go:34]: Was just a copy-paste comment, will remove.

File: cmd/snap-repair/runner.go
Comment (andrewphelpsj) [reply cmd/snap-repair/runner.go:1]: Removed the changes from snap-repair for a future PR.

File: overlord/configstate/configcore/store.go [hunk 14]
```diff
@@ -70,9 +70,14 @@ func handleStoreAccess(_ sysconfig.Device, cfg ConfGetter, _ *fsOnlyContext) err
 		return err
 	}
 
-	if err := os.MkdirAll(filepath.Dir(dirs.SnapRepairConfigFile), 0755); err != nil {
+	configFilePath := dirs.SnapRepairConfigFile
+	if opts != nil && opts.RootDir != "" {
+		configFilePath = dirs.SnapRepairConfigFileUnder(opts.RootDir)
+	}
```
Comment (pedronis) [inline overlord/configstate/configcore/store.go:75]: we want a test for this by using configcore.FilesystemOnlyApply as well and not just Run, if you look there should be examples in the preexisting test files

**Review by pedronis — COMMENTED** [review]
thank you, one final test comment

**Comment by andrewphelpsj** [comment]
Rebased to get unit test fixes from master branch.
